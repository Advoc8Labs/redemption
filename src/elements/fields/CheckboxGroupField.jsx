import React, { useMemo } from 'react';
import ReactForm from 'react-bootstrap/Form';
import { bootstrapStyles } from '../styles';
import { applyCheckboxStyles, composeCheckboxStyle } from './CheckboxField';

const applyCheckboxGroupStyles = (element, applyStyles) => {
  applyStyles.addTargets(['checkboxGroup']);
  applyStyles.applyHeight('checkboxGroup');
  applyStyles.applyWidth('checkboxGroup');

  return applyStyles;
};

function CheckboxGroupField({
  element,
  applyStyles,
  fieldLabel,
  fieldVal = [],
  otherVal = '',
  onChange = () => {},
  onOtherChange = () => {},
  onClick = () => {},
  elementProps = {}
}) {
  const servar = element.servar;
  const otherChecked = fieldVal.includes(otherVal);

  const styles = useMemo(() => {
    applyCheckboxStyles(element, applyStyles);
    applyCheckboxGroupStyles(element, applyStyles);

    return applyStyles;
  }, [applyStyles]);

  return (
    <div
      css={{
        display: 'flex',
        flexWrap: 'wrap',
        width: '100%',
        ...applyStyles.getTarget('fc')
      }}
      {...elementProps}
    >
      {fieldLabel}
      {servar.metadata.options.map((opt, i) => {
        return (
          <ReactForm.Check
            type='checkbox'
            id={`${servar.key}-${i}`}
            key={`${servar.key}-${i}`}
            name={opt}
            label={opt}
            checked={fieldVal.includes(opt)}
            onChange={onChange}
            onClick={onClick}
            style={{
              display: 'flex',
              alignItems: 'center',
              marginBottom: '5px'
            }}
            css={{
              ...composeCheckboxStyle(styles),
              ...styles.getTarget('checkboxGroup')
            }}
          />
        );
      })}
      {servar.metadata.other && (
        <div
          style={{
            display: 'flex',
            alignItems: 'center',
            ...styles.getTarget('checkboxGroup')
          }}
        >
          <ReactForm.Check
            type='checkbox'
            id={`${servar.key}-`}
            key={`${servar.key}-`}
            name={otherVal}
            label='Other'
            checked={otherChecked}
            onChange={onChange}
            onClick={onClick}
            style={{
              display: 'flex',
              alignItems: 'center'
            }}
            css={{
              ...composeCheckboxStyle(styles)
            }}
          />
          <ReactForm.Control
            type='text'
            css={{
              marginLeft: '5px',
              width: 'initial',
              ...bootstrapStyles,
              ...applyStyles.getTarget('field'),
              '&:focus': applyStyles.getTarget('active'),
              '&:hover': applyStyles.getTarget('hover')
            }}
            id={servar.key}
            value={otherVal || ''}
            onChange={onOtherChange}
            onClick={onClick}
            maxLength={servar.max_length}
            minLength={servar.min_length}
            required={otherChecked}
          />
        </div>
      )}
    </div>
  );
}

export default CheckboxGroupField;
