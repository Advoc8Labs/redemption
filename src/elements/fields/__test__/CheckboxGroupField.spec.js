import React from 'react';
import { create, act } from 'react-test-renderer';
import CheckboxGroupField from '../CheckboxGroupField';
import ApplyStyles from '../../styles';

describe('CheckboxGroupField', () => {
  it('renders a Radio Button group', () => {
    // Arrange
    const element = {
      id: 'fieldId',
      apply_styles: {},
      styles: {
        height: '50',
        height_unit: '%',
        width: '50',
        width_unit: 'px',
        background_color: '000000',
        uploader_padding_top: '0',
        uploader_padding_bottom: '0',
        uploader_padding_left: '0',
        uploader_padding_right: '0',
        cta_padding_top: '0',
        cta_padding_bottom: '0',
        cta_padding_left: '0',
        cta_padding_right: '0'
      },
      mobile_styles: {},
      servar: {
        key: 'key',
        name: 'Upload file...',
        metadata: {
          file_types: 'image/*',
          options: []
        },
        type: 'select'
      },
      properties: { icon_url: 'https://feathery.tech' }
    };
    const props = {
      element,
      fieldLabel: 'label',
      fieldVal: 'val',
      otherVal: 'otherVal',
      step: { servar_fields: [element] },
      fieldValues: { key: 'val' },
      updateFieldValues: { newKey: 'newVal' },
      onChange: jest.fn(),
      handleOtherStateChange: jest.fn(),
      onClick: jest.fn(),
      applyStyles: new ApplyStyles(element, [])
    };

    // Act
    let component;
    act(() => {
      component = create(<CheckboxGroupField {...props} />);
    });

    // Assert
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
