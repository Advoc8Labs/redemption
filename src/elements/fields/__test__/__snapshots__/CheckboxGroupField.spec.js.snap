// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxGroupField renders a Radio Button group 1`] = `
<div
  className="css-1ydmt8u-CheckboxGroupField"
>
  label
</div>
`;
